packages/core/src/codewhisperer/tracker/userWrittenCodeTracker.ts

```diff
@@ -0,0 +1,196 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import { getLogger } from '../../shared/logger/logger'
+import { runtimeLanguageContext } from '../util/runtimeLanguageContext'
+import { TelemetryHelper } from '../util/telemetryHelper'
+import { AuthUtil } from '../util/authUtil'
+import { getSelectedCustomization } from '../util/customizationUtil'
+import { codeWhispererClient as client } from '../client/codewhisperer'
+import { isAwsError } from '../../shared/errors'
+import { CodewhispererLanguage } from '../../shared'
+
+/**
+ * This singleton class is mainly used for calculating the user written code
+ * for active Amazon Q users.
+ * It reports the user written code per 5 minutes when the user is coding and using Amazon Q features
+ */
+export class UserWrittenCodeTracker {
```

Comment:
why is Q constantly adding new global singletones to track more state? why can't the existing one(s) be used?

Reply:
there are 4 different "trackers" https://github.com/aws/aws-toolkit-vscode/tree/32a4af5f1346d8be710c6b46202a181faca4c8d6/packages/core/src/codewhisperer/tracker

Reply:
This is a product requirement. 

The previous `CodeCoverageTracker.ts` needs to be kept as it is to provide data for the current enterprise user dashboard. In the current enterprise user dashboard, we display the % written by Q inline suggestion among all user written code. 

However, as more Q features are contributing code, we are shifting to a newer dashboard V2 that displays code contribution per Q feature. This needs to be running in shadow mode for sometime while the V1 dashboard needs to stay. Therefore, I need to create a different tracker just to provide data point for dashboard V2.

Reply:
The `codeWhispererTracker.ts` was meant to compute the edit % post inline completion acceptance. It is not meant to calculate the % code written. It tracks the position of snippets and how much it was modified.

Reply:
I do not have context of `lineTracker.ts`.

Reply:
Let me know if you want me to completely merge the UserWrittenCodeTracker.ts into CodeCoverageTracker.ts. 

IMO this is doable but it will add more if else and hard-wiring, making it even more error prone. Our expectation is to completely get rid of the old CodeCoverageTracker.ts in a few weeks.

Reply:
`lineTracker` is for UI purpose, so it's kinda different scope. 
But we do have 2 code percentage tracker(1 pre-existing and the one added here) and 1 modification tracker. IMO it would be better if we could merge % tracker into single one, but that will requires more time and planning.

Reply:
Lei do you know is service side ever using userModification?

Reply:
asking because given the fact that we hardly can track user's modification on client side, i don't really understand we bother keep this one

Reply:
agree with Justin that we better unify and merge these states machine otherwise it's error prone and hard to debug

Reply:
is there a way that we can define the message going to the %tracker with a field like `source` which could be `inline suggestion`, `dev` or `inline chat` etc.

Reply:
> Our expectation is to completely get rid of the old CodeCoverageTracker.ts in a few weeks.

That definitely helps a lot. Is there a tracking issue for that?

Reply:
>That definitely helps a lot. Is there a tracking issue for that?

I cannot post link here.